### ISSUE TITLE
Automatically get body from koa-bodyparser

### ISSUE THREAD

**Issue by watson** [elastic/apm-agent-nodejs#409]
If using the popular `koa-bodyparser` module, our `captureBody` config option currently doesn't have any effect. It would be nice if the agent natively supported extracting the body from this module. Either by getting it from `ctx.request.body` where it's stored by the `koa-bodyparser`, or by patching `koa-bodyparser` directly.

**Comment by watson** [comment]
We could instead consider getting the body directly from the request stream using https://github.com/watson/stream-observer

**Comment by Qard** [comment]
If we try to collect the request body with stream-observer we should make sure we have an escape condition decided for if the body gets too big.

**Comment by watson** [comment]
Yes. We already have [an upper limit on 2048 bytes](https://github.com/elastic/apm-agent-nodejs/blob/61f2ca4abdbd8c1452aea87bceeb0f8eae0968c2/lib/parsers.js#L14) after which the body is truncated. We should respect that and just stop listening.

The `stream-observer` module returns a function which you can call at any time to stop listening for data.

**Comment by Qard** [comment]
Ah, cool. Sounds like it might be a good option then. :)

**Comment by ilionic** [comment]
Not sure it's related - if using `koa-bodyparser` getting `GET unknown route`

**Comment by Qard** [comment]
`koa-bodyparser` wouldn't cause that, but you'll currently only get routing information if you use `koa-router`. Using `koa` directly, or using other routing modules will not record a different transaction name. You can use `agent.setTransactionName('custom name')` if you want to provide a name yourself.

**Comment by junchaozhang** [comment]
@watson  does this problem fix? 
If not fix, how to make the `koa-bodyparser` work?
I try the 
```
  captureBody: 'all',
``` 

I can not get the body information now.

**Comment by watson** [comment]
No, given the way `koa-bodyparser` work, we need to either patch it somehow or we need to hook in at a lower level. The way `captureBody` works currently is by expecting that the body is set as a property on the request object, either with the name `req.body`, `req.json` or `req.payload`. The `koa-bodyparser` module does not do that.

**Comment by junchaozhang** [comment]
> No, given the way `koa-bodyparser` work, we need to either patch it somehow or we need to hook in at a lower level. The way `captureBody` works currently is by expecting that the body is set as a property on the request object, either with the name `req.body`, `req.json` or `req.payload`. The `koa-bodyparser` module does not do that.

And do we have any plan to make it work?

**Comment by watson** [comment]
Yes, that's why I created this issue. But it's not yet scheduled when it will be implemented.

To help us prioritize, could you share a little information on what your use-case for capturing the body is? Is it during the development phase of the application, or in a production environment? How large are the avg. bodies you want to capture, what's their content-type etc.

**Comment by junchaozhang** [comment]
Here is a demo
```javascript
require('elastic-apm-node').start({
  // Override service name from package.json
  // Allowed characters: a-z, A-Z, 0-9, -, _, and space
  serviceName: 'helloworld',

  // Use if APM Server requires a token
  secretToken: '',

  // Set custom APM Server URL (default: http://localhost:8200)
  serverUrl: 'http://localhost:8200',
  logLevel: 'trace',
  asyncHooks: false,
  captureHeaders: true,
  captureBody: 'all',
  errorOnAbortedRequests: true,
  abortedErrorThreshold: true,
  transactionSampleRate: '1.0',

});

const Koa = require('koa');
const Router = require('koa-router');
const request = require('superagent');

const app = new Koa();
const router = new Router();

router.get('/tt', async (ctx, next) => {
  const response = await request.get('https://petstore.swagger.io/v2/store/inventory');
  ctx.body = response.body;
})

app.use(router.routes())
  .use(router.allowedMethods());

app.use(async ctx => {
  ctx.req.body = ctx.request.body
})

app.listen(3000);
```

**Comment by Qard** [comment]
@watson We should probably just capture the body from the `raw-body` module. As far as I know, _every_ body parser uses it.

**Comment by sibelius** [comment]
so no request body neither response body are recorded into apm?

**Comment by sibelius** [comment]
where is the best place to add a test to this ?
